File: workers/market_integrity_policy.py
"""Strict market-data integrity gates shared by alpha and settlement workers."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any


ALLOWED_MARKET_SOURCES = {
    "sportsbook_manual_verified",
    "bookmaker_manual_verified",
    "sportsdataio_manual_verified",
    "the_odds_api_manual_verified",
}


def _parse_timestamp(value: Any) -> datetime | None:
    if not value:
        return None
    if isinstance(value, datetime):
        parsed = value
    else:
        try:
            parsed = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
        except (TypeError, ValueError):
            return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc)


def _bool(value: Any) -> bool:
    if isinstance(value, bool):
        return value
    return str(value or "").strip().lower() in {"1", "true", "yes"}


def _common_reasons(snapshot: dict[str, Any]) -> list[str]:
    reasons: list[str] = []
    if not _bool(snapshot.get("canonical_match")):
        reasons.append("MATCH_NOT_CANONICAL")
    if _bool(snapshot.get("duplicate")):
        reasons.append("DUPLICATE_EXPOSURE")
    if str(snapshot.get("source_name") or "").strip().lower() not in ALLOWED_MARKET_SOURCES:
        reasons.append("SOURCE_NOT_ALLOWED")
    if not snapshot.get("evidence_id") or not snapshot.get("screenshot_sha256"):
        reasons.append("EVIDENCE_MISSING")
    return reasons


def _decision(status_ok: str, status_bad: str, reasons: list[str]) -> dict[str, Any]:
    unique = list(dict.fromkeys(reasons))
    return {
        "eligible": not unique,
        "status": status_ok if not unique else status_bad,
        "reasons": unique,
        "audit_only": bool(unique),
    }


def validate_entry_snapshot(snapshot: dict[str, Any]) -> dict[str, Any]:
    reasons = _common_reasons(snapshot)
    role = str(snapshot.get("snapshot_type") or "").strip().lower()
    if role not in {"entry", "current"}:
        reasons.append("ENTRY_ROLE_INVALID")
    if not _bool(snapshot.get("safe_for_entry")):
        reasons.append("SAFE_FOR_ENTRY_FALSE")
    captured = _parse_timestamp(snapshot.get("captured_at"))
    kickoff = _parse_timestamp(snapshot.get("kickoff"))
    if captured is None:
        reasons.append("INVALID_CAPTURE_TIMESTAMP")
    if kickoff is None:
        reasons.append("INVALID_KICKOFF_TIMESTAMP" if snapshot.get("kickoff") else "MISSING_KICKOFF")
    if captured is not None and kickoff is not None and captured >= kickoff:
        reasons.append("ENTRY_NOT_PREGAME")
    if str(snapshot.get("stale_status") or "").strip().lower() != "fresh":
        reasons.append("ENTRY_STALE")
    return _decision("ENTRY_VALID", "ENTRY_AUDIT_ONLY", reasons)


def closing_window_quality(captured_at: Any, kickoff_at: Any) -> str:
    captured = _parse_timestamp(captured_at)
    kickoff = _parse_timestamp(kickoff_at)
    if captured is None:
        return "INVALID_CAPTURE_TIMESTAMP"
    if kickoff is None:
        return "INVALID_KICKOFF_TIMESTAMP" if kickoff_at else "MISSING_KICKOFF"
    minutes_before = (kickoff - captured).total_seconds() / 60.0
    if 3.0 <= minutes_before <= 10.0:
        return "CAPTURED_ON_TIME"
    return "CAPTURED_TOO_EARLY" if minutes_before > 10.0 else "CAPTURED_LATE"


def validate_closing_snapshot(snapshot: dict[str, Any]) -> dict[str, Any]:
    reasons = _common_reasons(snapshot)
    if str(snapshot.get("snapshot_type") or "").strip().lower() != "closing":
        reasons.append("CLOSING_ROLE_INVALID")
    if not _bool(snapshot.get("safe_for_closing")):
        reasons.append("SAFE_FOR_CLOSING_FALSE")
    quality = closing_window_quality(snapshot.get("captured_at"), snapshot.get("kickoff"))
    if quality != "CAPTURED_ON_TIME":
        reasons.append(quality)
    decision = _decision("CLOSING_VALID", "CLOSING_AUDIT_ONLY", reasons)
    decision["closing_quality"] = quality
    return decision


def validate_settlement_eligibility(
    entry: dict[str, Any],
    closing: dict[str, Any],
    *,
    result_final: bool,
    result_source_verified: bool,
) -> dict[str, Any]:
    reasons = [f"ENTRY:{reason}" for reason in entry.get("reasons", [])]
    reasons.extend(f"CLOSING:{reason}" for reason in closing.get("reasons", []))
    if not entry.get("eligible"):
        reasons.append("ENTRY_CHAIN_INVALID")
    if not closing.get("eligible"):
        reasons.append("CLOSING_CHAIN_INVALID")
    if not result_final:
        reasons.append("RESULT_NOT_FINAL")
    if not result_source_verified:
        reasons.append("RESULT_SOURCE_NOT_VERIFIED")
    return _decision("READY_FOR_SETTLEMENT", "SETTLEMENT_BLOCKED", reasons)


def validate_clean_sample_eligibility(
    settlement: dict[str, Any], *, settlement_final: bool, clv_valid: bool
) -> dict[str, Any]:
    reasons = list(settlement.get("reasons", []))
    if not settlement.get("eligible"):
        reasons.append("SETTLEMENT_CHAIN_INVALID")
    if not settlement_final:
        reasons.append("SETTLEMENT_NOT_FINAL")
    if not clv_valid:
        reasons.append("CLV_NOT_VALID")
    return _decision("CLEAN_V2_ELIGIBLE", "LEGACY_OR_AUDIT_ONLY", reasons)

File: backend/src/trading/market-integrity-policy.ts
import { closingWindowDiagnostics } from "./timezone.js";

export const ALLOWED_MARKET_SOURCES = new Set([
  "sportsbook_manual_verified",
  "bookmaker_manual_verified",
  "sportsdataio_manual_verified",
  "the_odds_api_manual_verified"
]);

export type IntegrityDecision = {
  eligible: boolean;
  status: string;
  reasons: string[];
  audit_only: boolean;
};

export type MarketSnapshotIntegrityInput = {
  capturedAt?: string | null;
  kickoff?: string | null;
  sourceName?: string | null;
  evidenceId?: string | null;
  screenshotSha256?: string | null;
  snapshotType?: string | null;
  staleStatus?: string | null;
  safeForEntry?: boolean | null;
  safeForClosing?: boolean | null;
  canonicalMatch?: boolean | null;
  duplicate?: boolean | null;
};

export type SettlementIntegrityInput = {
  entry: IntegrityDecision;
  closing: IntegrityDecision;
  resultFinal?: boolean | null;
  resultSourceVerified?: boolean | null;
  settlementFinal?: boolean | null;
  clvValid?: boolean | null;
};

function normalize(value: string | null | undefined) {
  return String(value || "").trim().toLowerCase();
}

export function normalizeSport(value: string | null | undefined) {
  const sport = normalize(value);
  if (["soccer", "football", "futbol", "fútbol"].includes(sport)) return "soccer";
  if (["baseball", "baseball/mlb", "mlb"].includes(sport)) return "baseball";
  if (["basketball", "nba"].includes(sport)) return "basketball";
  return sport || "unknown";
}

export function isCanonicalMatch(input: Pick<MarketSnapshotIntegrityInput, "canonicalMatch" | "duplicate">) {
  return input.canonicalMatch === true && input.duplicate !== true;
}

function hasEvidence(input: MarketSnapshotIntegrityInput) {
  return Boolean(input.evidenceId && String(input.evidenceId).trim())
    && Boolean(input.screenshotSha256 && String(input.screenshotSha256).trim());
}

function commonReasons(input: MarketSnapshotIntegrityInput) {
  const reasons: string[] = [];
  if (input.canonicalMatch !== true) reasons.push("MATCH_NOT_CANONICAL");
  if (input.duplicate === true) reasons.push("DUPLICATE_EXPOSURE");
  if (!ALLOWED_MARKET_SOURCES.has(normalize(input.sourceName))) reasons.push("SOURCE_NOT_ALLOWED");
  if (!hasEvidence(input)) reasons.push("EVIDENCE_MISSING");
  return reasons;
}

function validTimestamp(value: string | null | undefined) {
  if (!value) return false;
  return !Number.isNaN(new Date(value).getTime());
}

export function validateEntrySnapshot(input: MarketSnapshotIntegrityInput): IntegrityDecision {
  const reasons = commonReasons(input);
  const role = normalize(input.snapshotType);
  if (role !== "entry" && role !== "current") reasons.push("ENTRY_ROLE_INVALID");
  if (input.safeForEntry !== true) reasons.push("SAFE_FOR_ENTRY_FALSE");
  if (!validTimestamp(input.capturedAt)) reasons.push("INVALID_CAPTURE_TIMESTAMP");
  if (!validTimestamp(input.kickoff)) reasons.push(input.kickoff ? "INVALID_KICKOFF_TIMESTAMP" : "MISSING_KICKOFF");

  if (validTimestamp(input.capturedAt) && validTimestamp(input.kickoff)) {
    if (new Date(input.capturedAt!).getTime() >= new Date(input.kickoff!).getTime()) {
      reasons.push("ENTRY_NOT_PREGAME");
    }
  }
  if (normalize(input.staleStatus) !== "fresh") reasons.push("ENTRY_STALE");

  return {
    eligible: reasons.length === 0,
    status: reasons.length === 0 ? "ENTRY_VALID" : "ENTRY_AUDIT_ONLY",
    reasons,
    audit_only: reasons.length > 0
  };
}

export function validateClosingSnapshot(input: MarketSnapshotIntegrityInput): IntegrityDecision & {
  closing_quality: string;
  closing_window_start: string | null;
  closing_window_end: string | null;
} {
  const reasons = commonReasons(input);
  if (normalize(input.snapshotType) !== "closing") reasons.push("CLOSING_ROLE_INVALID");
  if (input.safeForClosing !== true) reasons.push("SAFE_FOR_CLOSING_FALSE");

  const diagnostics = computeClosingQuality(input.capturedAt, input.kickoff);
  if (diagnostics.closing_quality !== "CAPTURED_ON_TIME") reasons.push(diagnostics.closing_quality);

  return {
    eligible: reasons.length === 0,
    status: reasons.length === 0 ? "CLOSING_VALID" : "CLOSING_AUDIT_ONLY",
    reasons,
    audit_only: reasons.length > 0,
    closing_quality: diagnostics.closing_quality,
    closing_window_start: diagnostics.closing_window_start,
    closing_window_end: diagnostics.closing_window_end
  };
}

export function computeClosingQuality(capturedAt?: string | null, kickoff?: string | null) {
  return closingWindowDiagnostics(String(capturedAt || ""), kickoff);
}

export function validateSettlementEligibility(input: SettlementIntegrityInput): IntegrityDecision {
  const reasons = [
    ...input.entry.reasons.map((reason) => `ENTRY:${reason}`),
    ...input.closing.reasons.map((reason) => `CLOSING:${reason}`)
  ];
  if (!input.entry.eligible) reasons.push("ENTRY_CHAIN_INVALID");
  if (!input.closing.eligible) reasons.push("CLOSING_CHAIN_INVALID");
  if (input.resultFinal !== true) reasons.push("RESULT_NOT_FINAL");
  if (input.resultSourceVerified !== true) reasons.push("RESULT_SOURCE_NOT_VERIFIED");

  return {
    eligible: reasons.length === 0,
    status: reasons.length === 0 ? "READY_FOR_SETTLEMENT" : "SETTLEMENT_BLOCKED",
    reasons: [...new Set(reasons)],
    audit_only: reasons.length > 0
  };
}

export function validateCleanSampleEligibility(input: SettlementIntegrityInput): IntegrityDecision {
  const settlement = validateSettlementEligibility(input);
  const reasons = [...settlement.reasons];
  if (input.settlementFinal !== true) reasons.push("SETTLEMENT_NOT_FINAL");
  if (input.clvValid !== true) reasons.push("CLV_NOT_VALID");
  return {
    eligible: reasons.length === 0,
    status: reasons.length === 0 ? "CLEAN_V2_ELIGIBLE" : "LEGACY_OR_AUDIT_ONLY",
    reasons: [...new Set(reasons)],
    audit_only: reasons.length > 0
  };
}

export function computeClvEligibility(input: SettlementIntegrityInput): IntegrityDecision {
  const settlement = validateSettlementEligibility(input);
  const reasons = [...settlement.reasons];
  if (input.settlementFinal !== true) reasons.push("SETTLEMENT_NOT_FINAL");
  return {
    eligible: reasons.length === 0,
    status: reasons.length === 0 ? "CLV_ELIGIBLE" : "LEGACY_CLV_PREVIEW_ONLY",
    reasons: [...new Set(reasons)],
    audit_only: reasons.length > 0
  };
}

export function isCleanV2Eligible(input: SettlementIntegrityInput) {
  return validateCleanSampleEligibility(input).eligible;
}

export function isAuditOnly(decision: IntegrityDecision) {
  return decision.audit_only;
}

export function isLegacyOnly(input: SettlementIntegrityInput) {
  return !isCleanV2Eligible(input);
}
